### PR TITLE
feat: report customer volume from attributed trades

### PR DESCRIPTION
**Pull request by findolor**
## Chained PR

- Requires #160 (`feat: embed API key attribution in swap calldata`) to merge first.
- This is the only reporting PR in the stack.

## Motivation

#160 makes generated swap calldata attributable, but we still need to determine which signed calls were actually executed and how much customer volume they produced. Generated quotes and unused calldata must not count.

The reporting data must remain private, survive delayed execution and API-key or signer rotation, and start from an explicit rollout block instead of indexing historical volume from genesis.

## Solution

- Run a single-replica Rocket background worker every 60 seconds, starting at Base block `48963501` and processing at most 250 indexed trades per target per interval.
- Read confirmed `TakeOrderV3` trades and emitted signed contexts from the existing Raindex SQLite database. This avoids redundant RPC transaction fetches because the indexed event already contains the canonical signer, signature, context, transaction, order, taker, tokens, and executed amounts.
- Read each Raindex watermark and batch in one SQLite snapshot, then advance a stable `(block_number, log_index, trade_id)` cursor atomically with attributed-trade writes.
- Verify the signed versioned frame against the trusted REST signer, event taker, and exact order hash before accepting the API-key hash.
- Preserve previously observed signer addresses and API-key hash-to-customer tombstones so delayed executions remain attributable after signer rotation or hard key deletion. No generated calldata is stored.
- Store only confirmed attributed executions, exact Rain float amounts, customer identity snapshots, and sync cursors in the private application database.
- Add admin-only reporting endpoints:
  - `GET /admin/attribution/executions` with block, API-key hash, transaction hash, limit, and stable keyset pagination filters.
  - `GET /admin/attribution/volume` grouped exactly by API-key hash, chain, and token pair.
- Reset and safely rescan derived attribution data when the configured start block changes.
- Join the worker cleanly during Rocket shutdown and keep OpenAPI parameter names aligned with the camel-case REST contract.

## Privacy and deployment

- Uses the existing `ST0X_GATING_SIGNER_KEY`; no signer environment variable is added.
- RPC URLs and credentials remain in the existing private registry configuration.
- The API key secret and signer private key are never stored in reporting rows or returned by the endpoints.
- Preview and production are configured to start at block `48963501`, which includes the known staging execution.
- Service-only deployments compare the installed attribution settings with the checked-out environment config as well as checking the signer unit. Missing or stale settings automatically trigger the required system deployment before the service deploy.

## Regression coverage

- Verifies attribution signature recovery and the exact signed-context layout shared with #160.
- Processes the known staging transaction fixture (`0x48f6ed8a67769c007491262e72b78eb934a0a53bc0d67506081fc9a1c35c276f`) at block `48963501`, including the `0.001` input and `0.205184` output amounts.
- Covers idempotent retries, empty-tail checkpoints, exact-full and partial batches, shared block/log positions, start-block changes, signer rotation, atomic API-key tombstones, exact volume aggregation, admin authorization, transaction filtering, stable pagination, and OpenAPI query names.

## Checks

- `nix develop -c cargo fmt --all`
- `nix develop -c rainix-rs-static`
- `nix develop -c cargo test` (311 passed)
- `nix develop -c git diff --check`
- Three-angle simplification review; actionable findings fixed.
- Two-reviewer staged local review plus one focused second pass; actionable findings fixed and retested.

## Preview verification

- Preview service deployment: https://github.com/ST0x-Technology/st0x.rest.api/actions/runs/29917111920
- Preview system-config deployment: https://github.com/ST0x-Technology/st0x.rest.api/actions/runs/29918959899
- The live worker registered signer `0x385817222DBb0Ee011B8c1722479C2A462b4cE2E` and advanced from start block `48963501` to the current Raindex watermark.
- `GET /admin/attribution/executions` returned the exact known transaction `0x48f6ed8a67769c007491262e72b78eb934a0a53bc0d67506081fc9a1c35c276f`, API-key hash `0xee79738dcdfded041361028dcf0162b49eafa0a2fed61ba63d8fe275253aa8dc`, label `bruno-preview`, and owner `Arda`.
- `GET /admin/attribution/volume` returned one trade with total input `0.001` and total output `0.2051844466600199401794616151545363908275174476570289132602193419741`.
- A short-lived admin credential was used only for the live endpoint checks, then deleted; both its API-key row and empty attribution-directory row were verified absent.

## Summary by CodeRabbit

* **New Features**
  * Added attribution reporting worker for confirmed trades with periodic, cursor-based synchronization and signed-context validation.
  * Added admin endpoints for querying attributed executions and aggregated attribution volume with stable filtering/pagination.
  * Added persistence for attributed trades, signers, attribution API key metadata, and sync cursors.
  * Added attribution configuration for start block, sync interval, and batch size (with safer defaults).

* **Bug Fixes**
  * Improved deployment gating to verify both service prerequisites and that remote attribution configuration matches expected values.
  * Ensured attribution identity snapshots are preserved during API key create/delete.

* **Chores**
  * Added SQL migrations and supporting indexes for the new attribution storage and performance.

### CODE REVIEW

**Comment by findolor** [comment]
* **#162** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.rest.api/162?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#161** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.rest.api/161?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.rest.api/161?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#160** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.rest.api/160?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=161).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by 0xgleb — COMMENTED** [review]
agent misfire

File: src/attribution_reporting.rs [hunk 1]
```diff
@@ -0,0 +1,1058 @@
...
+        has_cursor,
+        last_block,
+        last_log_index,
+        &last_trade_id,
+        batch_size,
+    )
+    .await?;
+    let trades = group_context_rows(rows);
+    let next_cursor = if trades.len() < batch_size as usize {
+        (target.last_indexed_block, i64::MAX, String::new())
```
Comment (0xgleb) [inline src/attribution_reporting.rs:454]: minor: document and test the watermark-completeness invariant behind advancing a partial or empty batch to block MAX. Without that upstream guarantee, late rows at the watermark block are skipped.

File: src/attribution_reporting.rs [hunk 2]
```diff
@@ -0,0 +1,1058 @@
...
+    trade: &IndexedTrade,
+    trusted_signers: &[Address],
+) -> Option<B256> {
+    let (signer, signature) = parse_signer_and_signature(&candidate.encoded_signer_and_signature)?;
+    let context = candidate
+        .values
+        .iter()
+        .map(|value| B256::from_str(value).ok())
+        .collect::<Option<Vec<_>>>()?;
+    let taker = Address::from_str(&trade.transaction_sender).ok()?;
```
Comment (0xgleb) [inline src/attribution_reporting.rs:707]: minor: confirm that transaction_sender is the intended signed taker for every supported execution path. Relayed executions are dropped when sender and taker differ.

File: src/attribution_reporting.rs [hunk 3]
```diff
@@ -0,0 +1,1058 @@
...
+async fn process_target(
+    app_pool: &DbPool,
+    source_transaction: &mut sqlx::Transaction<'_, sqlx::Sqlite>,
+    trusted_signers: &[Address],
+    configured_start_block: i64,
+    batch_size: u32,
+    target: &SyncTarget,
+    identities: &HashMap<B256, ApiKeyIdentity>,
+) -> Result<u64, AttributionReportingError> {
+    let mut stored_cursor = sqlx::query_as::<_, CursorRow>(
```
Comment (0xgleb) [inline src/attribution_reporting.rs:412]: minor: keep the cursor read and update transactional, or explicitly enforce the single-worker invariant. Overlapping workers can otherwise regress progress.

File: src/attribution_reporting.rs [hunk 4]
```diff
@@ -0,0 +1,1058 @@
...
+            tracing::error!(%error, "failed to record current attribution signer at startup");
+        }
+        let worker = self.clone();
+        let mut shutdown = rocket.shutdown();
+        let handle = tokio::spawn(async move {
+            let mut interval = tokio::time::interval(worker.interval);
+            interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+            loop {
+                tokio::select! {
```
Comment (0xgleb) [inline src/attribution_reporting.rs:81]: minor: make the in-flight sync cancellation-aware or bound shutdown waiting; shutdown is observed only before processing starts.

File: src/routes/attribution_admin.rs [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+) -> Result<Vec<AttributionVolume>, ApiError> {
+    let start_block = optional_sqlite_integer(start_block, "startBlock")?;
+    let end_block = optional_sqlite_integer(end_block, "endBlock")?;
+    let mut query = QueryBuilder::<Sqlite>::new(
+        "SELECT chain_id, api_key_hash, input_token, input_amount, output_token, output_amount \
+         FROM attributed_trades WHERE 1 = 1",
+    );
+    push_filters(&mut query, api_key_hash, start_block, end_block, None);
+    let mut rows = query.build_query_as::<AttributionVolumeRow>().fetch(pool);
+    let mut groups: BTreeMap<VolumeKey, VolumeAccumulator> = BTreeMap::new();
```
Comment (0xgleb) [inline src/routes/attribution_admin.rs]: minor: add a bounded aggregation strategy or explicit operational limit. This endpoint retains all distinct groups and identity rows in memory.

File: src/routes/attribution_admin.rs [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+    .map_err(query_error)?;
+    let identities: HashMap<String, AttributionIdentityRow> = identity_rows
+        .into_iter()
+        .map(|identity| (identity.api_key_hash.clone(), identity))
+        .collect();
+
+    groups
+        .into_iter()
+        .map(|(key, value)| {
+            let identity = identities.get(&key.api_key_hash);
```
Comment (0xgleb) [inline src/routes/attribution_admin.rs]: minor: document that volume uses current identity metadata, or use execution snapshots for consistent historical labeling.

**Review by 0xgleb — CHANGES_REQUESTED** [review]

**Comment by findolor** [comment]
@codex Please perform a full code review of this PR.

Review goals:
- Follow all applicable AGENTS.md instructions, including any nested AGENTS.md files that apply to changed files.
- Inspect repository convention files, contribution docs, lint/config rules, and nearby existing code patterns before judging the implementation.
- Verify the PR follows the repo's established architecture, naming, error handling, testing, styling, and abstraction conventions.
- Prioritize actionable findings: bugs, regressions, security issues, data-loss risks, race conditions, edge cases, missing validation, broken API contracts, and insufficient tests.
- Check whether the code introduces unnecessary duplication, weak abstractions, over-engineering, or behavior that should reuse existing helpers/components.
- Look for security and privacy issues, including unsafe input handling, auth/permission gaps, secret exposure, injection risks, SSRF/path traversal/XSS/CSRF where relevant, and overly broad logging.
- Review tests for meaningful coverage of the changed behavior, including failure paths and important edge cases.
- Call out compatibility or migration risks if schemas, APIs, configs, background jobs, or user-facing behavior changed.

Please structure the review like this:
1. Findings first, ordered by severity.
2. For each finding, include file/line references, the concrete risk, and a suggested fix.
3. If there are no blocking issues, say that explicitly.
4. Include any open questions or assumptions.
5. Keep summary brief and avoid restating the diff.

Do not leave style-only comments unless they conflict with repository conventions or materially affect maintainability.

**Review by JuaniRios — APPROVED** [review]

**Review by 0xgleb — APPROVED** [review]
